Improved Kover documentation (#369)
Kotlin/kotlinx-kover: docs/gradle-plugin/index.md

## Table of contents

- [Features](#features)
- [Quickstart](#quickstart)
- [Using Kover tasks](#using-kover-tasks)
  - [Kotlin JVM project](#kotlin-jvm-project)
  - [Kotlin Android project](#kotlin-android-project)
  - [Kotlin multiplatform project](#kotlin-multiplatform-project)
- [Multiproject build](#multiproject-build)
  - [Single report over several projects](#single-report-over-several-projects)
- [Filtering reports](#filtering-reports)
- [Kover configuration](configuring)
- [Kover Tasks](#kover-tasks)
  - [Kover default tasks](#kover-default-tasks)
  - [Kover Android tasks](#kover-android-tasks)
- [Instrumentation](#instrumentation)
- [Using JaCoCo](#using-jacoco)
- [Implicit plugin dependencies](#implicit-plugin-dependencies)
- [Gradle Plugin DSL docs](dokka)

## Features

* Collection of code coverage through `JVM` test tasks.
* `HTML` and `XML` reports.
* Support for `Kotlin/JVM`, `Kotlin Multiplatform` and mixed `Kotlin-Java` sources with zero additional configuration.
* Support for `Kotlin Android` without the need to divide it into build types and flavours.
* Verification rules with bounds to keep track of coverage.

> * Customizable filters for instrumented classes.

Reviewer: Should this feature list be identical with one in the README?

Author: OK

    }
}
```

### Kotlin multiplatform project
These are projects that use Kotlin Multiplatfrom Gradle plugin (Kotlin MPP).
Working with such projects is identical to working with JVM and Android projects simultaneously, which are described above.

Both JVM and Android targets may be present in such projects, depending on the declared targets, it is necessary to use the appropriate Kover tasks.

If only targets with the JVM type are declared, then the [Kover default tasks](#kover-default-tasks) should be used.

If only a target with the Android type is declared then the [Kover android tasks](#kover-android-tasks) should be used.
In this case, starting default task like `koverHtmlReport` will result in a message `Task 'koverHtmlReport' will be skipped because no tests were executed` in the build log, and the report will not be generated.

A special case is when JVM and Android targets are present in the project at the same time -
in this case, Kover default tasks (e.g. `koverHtmlReport`) will generate reports for all JVM targets, and Kover Android tasks (e.g. `koverHtmlReportRelease`) generate reports for specific build variant.

> However, if there is a need for a single report to contain measurements for both JVM targets and for any Android build variant, in this case, [reports merging](configuring#merging-reports) is used.

Reviewer: ```suggestion
However, if there is a need for a single report to contain measurements for both JVM targets and for any Android build variant, [reports merging](configuring#merging-reports) can be used.
```

> However, if there is a need for a single report to contain measurements for both JVM targets and for any Android build variant, in this case, [reports merging](configuring#merging-reports) is used.

Reviewer: ```suggestion
However, if there is a need for a single report to contain measurements for both JVM targets and for any Android build variant, [reports merging](configuring#merging-reports) can be used.
```

            classes("com.example.Class1", "com.example.Class3")
        }
    }
}
```
If inclusion and exclusion rules are specified at the same time, then excludes have priority over includes.
This means that even if a class is specified in both the inclusion and exclusion rules, it will be excluded from the report (e.g. class `com.example.Class1` above).

[Wildcards](configuring#class-name-with-wildcards) `*` and `?` are allowed in class names.

## Kover configuration

## Kover Tasks
### Kover default tasks
Default Kover tasks - are tasks that are always created when applying the Kover Gradle plugin.

These tasks generate a report for Kotlin/JVM or Java sources, Android source codes are not included in these reports by default.

Running Kover default tasks cause compilation of all Kotlin JVM and Java classes, and also run the corresponding tests for JVM targets.

Kover default task list:
- `koverHtmlReport` - Generate HTML report for Kotlin/JVM or Java classes
- `koverXmlReport` - Generate XML report for Kotlin/JVM or Java classes
- `koverVerify` - Verifies code coverage metrics of Kotlin/JVM or Java classes based on configured rules

Example:
```

> gradlew koverHtmlReport

Reviewer: I think it should be `./gradlew` for easier copy-pasting

In this case, it is necessary to disable the instrumentation of the problem class. As a side effect, the coverage of such a class will not be measured, and if the excluded class was declared in the project, then its coverage in the report will always be 0.
```kotlin
kover {
    excludeInstrumentation {
        classes("com.example.ExcludedClass")
    }
}
```

[Wildcards](configuring#class-name-with-wildcards) `*` and `?` are allowed in class names.

Typical error messages encountered with instrumentation problems:
```
No instrumentation registered! Must run under a registering instrumentation.
```

```
java.lang.VerifyError
	at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses0(Native Method)
```

## Using JaCoCo
Kover Gradle plugin provides the ability to use the [JaCoCo Coverage Library](https://github.com/jacoco/jacoco) to collect class coverage and generate reports.
However, full feature compatibility is not guaranteed.

To use JaCoCo with the default version in the project, specify
```kotlin
kover {
    useJacoco()
}
```
for the custom version
```kotlin
kover {
    useJacoco("0.8.8")
}
```

> **It is important that if used [dependencies in multi-project builds](#single-report-over-several-projects), one type of coverage library: embedded Kover or JaCoCo of the same version is used**

Reviewer: ```suggestion
**It is important that if [dependencies in multi-project builds](#single-report-over-several-projects) are specified, only one type of coverage library (embedded Kover or JaCoCo) is used across these projects**
```
